File: 프로그래머스/3/42898. 등굣길/등굣길.java
import java.util.*;
class Solution {
    public int solution(int m, int n, int[][] puddles) {
        int mod = 1000000007;
        int answer = 0;
        int [][] dp = new int[n+1][m+1];
        boolean[][] npuddles = new boolean[n+1][m+1];
        for(int[] puddle: puddles){
            npuddles[puddle[1]][puddle[0]] = true;
        }
        dp[1][1] = 1;
        for (int i =0; i<n+1; i++){
            for(int j = 0; j<m+1; j++){
                if(i == 1 && j == 1) continue;
                if(npuddles[i][j]){
                    dp[i][j] = 0;
                } else {
                    int f = (i > 0) ? dp[i-1][j] : 0;
                    int r = (j > 0) ? dp[i][j-1] : 0;
                    dp[i][j] = (f + r) % mod;
                }
            }
        }
        answer = dp[n][m] ;
        return answer;
    }
}

File: 프로그래머스/3/42898. 등굣길/등굣길.py
def solution(m, n, puddles):
    answer = 0
    dp = [[0]*(m+1) for i in range(n+1)]    
    dp[1][1] = 1
    # right, down 
    # 격자에 가능한 경로 개수 넣기
    # 주변 격자의 값 (left, up)
    puddles = [(x,y) for (y,x) in puddles]
    for i in range(1, n+1): # row
        for j in range(1,m+1): # col
            if i == 1 and j == 1:
                continue
            if (i,j) in puddles:
                dp[i][j] = 0
            else:
                dp[i][j] = (dp[i-1][j] + dp[i][j-1])
    
    answer = dp[n][m] % 1000000007
    return answer